src/protocols/routing/mod.rs

```diff
@@ -0,0 +1,270 @@
+
+pub(crate) const FORWARD_MSG_TYPE: &str = "https://didcomm.org/routing/2.0/forward";
+
+async fn find_did_comm_service<'dr>(
+    did: &str,
+    service_id: Option<&str>,
+    did_resolver: &'dr (dyn DIDResolver + 'dr),
+) -> Result<Option<Service>> {
+    let did_doc = did_resolver
+        .resolve(did)
+        .await
+        .context("Unable resolve DID")?
+        .ok_or_else(|| err_msg(ErrorKind::DIDNotResolved, "DID not found"))?;
+
+    match service_id {
+        Some(service_id) => {
+            let service: &Service = did_doc
+                .services
+                .iter()
+                .find(|&service| service.id == service_id)
+                .ok_or_else(|| {
+                    err_msg(
+                        ErrorKind::InvalidState,
+                        "Service with the specified ID not found",
+                    )
+                })?;
+
+            if let ServiceKind::DIDCommMessaging(_) = service.kind {
+                Ok(Some(service.clone()))
+            } else {
+                Err(err_msg(
+                    ErrorKind::InvalidState,
+                    "Service with the specified ID is not of DIDCommMessaging type",
+                ))
+            }
+        }
+
+        None => Ok(did_doc.services.iter().find_map(|service| {
+            if let ServiceKind::DIDCommMessaging(_) = service.kind {
+                Some(service.clone())
+            } else {
+                None
+            }
+        })),
+    }
+}
+
+fn unwrap_did_comm_service(service: &Service) -> Result<&DIDCommMessagingService> {
+    match service.kind {
+        ServiceKind::DIDCommMessaging(ref did_comm_service) => Ok(did_comm_service),
+        ServiceKind::Other(_) => Err(err_msg(
+            ErrorKind::InvalidState,
+            "Service is not of DIDCommMessaging type",
+        )),
+    }
+}
+
+async fn resolve_did_comm_services_chain<'dr>(
+    to: &str,
+    service_id: Option<&str>,
+    did_resolver: &'dr (dyn DIDResolver + 'dr),
+) -> Result<Vec<Service>> {
+    let (to_did, _) = did_or_url(to);
+
+    let service = find_did_comm_service(to_did, service_id, did_resolver).await?;
+
+    if service.is_none() {
+        return Ok(vec![]);
+    }
+
+    let mut service = service.unwrap();
+
+    let mut services = vec![service.clone()];
+    let mut service_endpoint = &unwrap_did_comm_service(&service)?.service_endpoint;
+
+    while is_did(service_endpoint) {
+        // Now alternative endpoints recursion is not supported
+        // (because it should not be used according to the specification)
+        if services.len() > 1 {
+            return Err(err_msg(
+                ErrorKind::InvalidState,
+                "DID doc defines alternative endpoints recursively",
+            ));
+        }
+
+        service = find_did_comm_service(service_endpoint, None, did_resolver)
+            .await?
+            .ok_or_else(|| {
+                err_msg(
+                    ErrorKind::InvalidState,
```

Comment:
Is it really an error? Probably we can just return an empty list in this case...
Even if we consider this as an error, I'm not sure we should raise InvalidState. Probably a specific error type 'InvalidDIDDoc`?